### PR TITLE
Set stored kcal to healthy kcal for testing

### PR DESCRIPTION
**Pull request by actual-nh**
#### Summary

Infrastructure "Set stored kcal to healthy kcal for testing"

#### Purpose of change

Randomizing test orders (#46473) reveals that one type of problematic interference is due to variable stored calories from prior activity.

#### Describe the solution

In `clear_character()` (in tests/player_helpers.cpp), set stored kcal to healthy kcal. (Note: Setting to any less causes errors; setting to any more puts BMI above 25.)

#### Testing

See #46473. (From that, some additional setting of vitamins may **also** be needed, but this is a start.)

#### Additional context

~~See #46934 for another case where it appears likely this will help.~~ Never mind on that one; a more frequent symptom is the "test subject" starving to death in the middle of a test...

### CODE REVIEW

**Comment by BrettDong** [comment]
Probably you need to update numbers in unit tests:
```
../tests/weary_test.cpp:170: FAILED:
1371   CHECK( info.transition_minutes( 0, 1, 130_minutes ) == Approx( 130 ).margin( 5 ) )
1372 with expansion:
1373   120 == Approx( 130.0 )
1374 with messages:
1375   Weariness: 3 Max Full Exert: EXTRA_EXERCISE Mult: 1
1376   BMR: 1741 Intake: 932 Tracker: 7 Thresh: 846 At: 0
1377   Cal: 55393/55000 Fatigue: 289 Morale: 0 Wgt: 76825 (BMI 25.1)
1378   Transition: Weariness lvl from 0 to 1 at 120 min (W 1024 Th 1020)
```

**Comment by actual-nh** [comment]
I am not sure why the Github build tests are still showing up as "in progress" - the g++-7 one had test failures, and the others were accordingly cancelled.

> Probably you need to update the numbers in unit tests:

Looks like it. The debug_weary_info output at the start indicates the prior test had formerly left the calories at over 55000, influencing the test; now, it isn't (at least via calories). That the results differed between g++-7 and the Travis+Appveyor is interesting, and may imply it would be preferable to set those differing by 10 between actual and expected to the mid-point.

> ```
> ../tests/weary_test.cpp:170: FAILED:
> 1371   CHECK( info.transition_minutes( 0, 1, 130_minutes ) == Approx( 130 ).margin( 5 ) )
> 1372 with expansion:
> 1373   120 == Approx( 130.0 )

> 1377   Cal: 55393/55000 Fatigue: 289 Morale: 0 Wgt: 76825 (BMI 25.1)
> 1378   Transition: Weariness lvl from 0 to 1 at 120 min (W 1024 Th 1020)
> ```

**Comment by actual-nh** [comment]
Interesting. The earlier build had given differing results between the General (g++-7) and Appveyor (and Travis - don't know yet on this one for Travis; seems to be delayed) for the 24-hour weary digging task - different enough that I couldn't set the expected test results to cover both of them:

Previous Travis:
```
  Transition: Weariness lvl from 0 to 1 at 120 min (W 1024 Th 1020)
  Transition: Weariness lvl from 1 to 2 at 250 min (W 2020 Th 1000)
  Transition: Weariness lvl from 2 to 3 at 360 min (W 2735 Th 983)
  Transition: Weariness lvl from 3 to 4 at 470 min (W 3228 Th 968)
  Transition: Weariness lvl from 4 to 5 at 590 min (W 3558 Th 951)
  Transition: Weariness lvl from 5 to 6 at 715 min (W 3805 Th 932)
  Transition: Weariness lvl from 6 to 7 at 810 min (W 3945 Th 920)
  Transition: Weariness lvl from 7 to 8 at 885 min (W 4059 Th 910)
```

Previous Appveyor:
```
  Transition: Weariness lvl from 0 to 1 at 120 min (W 1027 Th 1020)
  Transition: Weariness lvl from 1 to 2 at 250 min (W 2023 Th 1000)
  Transition: Weariness lvl from 2 to 3 at 360 min (W 2739 Th 983)
  Transition: Weariness lvl from 3 to 4 at 465 min (W 3207 Th 968)
  Transition: Weariness lvl from 4 to 5 at 590 min (W 3549 Th 951)
  Transition: Weariness lvl from 5 to 6 at 715 min (W 3795 Th 933)
  Transition: Weariness lvl from 6 to 7 at 815 min (W 3941 Th 919)
  Transition: Weariness lvl from 7 to 8 at 890 min (W 4056 Th 909)
```

Previous General (g++7):
```
  Transition: Weariness lvl from 0 to 1 at 125 min (W 1078 Th 1019)
  Transition: Weariness lvl from 1 to 2 at 250 min (W 2023 Th 1000)
  Transition: Weariness lvl from 2 to 3 at 360 min (W 2737 Th 983)
  Transition: Weariness lvl from 3 to 4 at 470 min (W 3208 Th 968)
  Transition: Weariness lvl from 4 to 5 at 600 min (W 3574 Th 949)
  Transition: Weariness lvl from 5 to 6 at 735 min (W 3765 Th 930)
  Transition: Weariness lvl from 6 to 7 at 840 min (W 3928 Th 916)
  Transition: Weariness lvl from 7 to 8 at 925 min (W 4037 Th 905)
```

Now, Appveyor (and General g++-7) match(es) the General (g++-7) one above. I think I'll try using sufficiently close to the General/g++-7 numbers above.

**Comment by BrettDong** [comment]
What is the reason do you think that the results differ? Is it because some unit test elsewhere corrupted internal state, or is it related to compiler / platform?

**Comment by actual-nh** [comment]
> What is the reason do you think that the results differ? Is it because some unit test elsewhere corrupted internal state, or is it related to compiler / platform?

Well, I can say from running them locally with just the weary tests that the results differ yet again, so it's at least partially the former.  Appveyor and the others differ in whether tests are run in parallel, so it's difficult to tell there. But the Github and Travis runs should not be differing in test order, so there is at least an intermittent compiler/platform issue.

**Comment by BrettDong** [comment]
Does it give the exact same result if run unit tests with the same RNG seed and on the exact same environment, say locally, multiple times?

**Comment by actual-nh** [comment]
> Does it give the exact same result if run unit tests with the same RNG seed and on the exact same environment, say locally, multiple times?

Actually, running just the weary tests 3 times locally, with different orderings (not much reordering for 3 tests, admittedly...) and different RNG seeds, gives consistent results - different from any of the above. I'll try doing a run of the full set of tests locally, to see if those differ, but it will take a while.

**Comment by actual-nh** [comment]
And now the Github g++-7 one is giving differing results on one of the two weary_recovery tasks...

Note that one limitation of running all tests is that there doesn't seem a way to do `--success` for a subset of tests in Catch - so I'm either only going to see where tests fail, or will need to collect lots and lots of output (sigh...). Oh. It would help a bit to just do the subset of tests that's currently in the same process in build.sh; will try that.

**Comment by actual-nh** [comment]
This one has all non-slow tests:
[all_test1_log.txt.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/5898638/all_test1_log.txt.gz)
Extract: 
[all_test1_weary_log_extract.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898648/all_test1_weary_log_extract.txt)

This one has just the weary tests, run with randomized order 4 times (2 before any changing of test targets, 2 after most recent update of test targets):
[all_weary_tests_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898644/all_weary_tests_log.txt)

I'm going to try doing just the weary tests 3 times, in the default order (or the ordering in my test list file -  need to check which), to see if the RNG seed is what's causing different results in the second 2 of the above.

**Comment by actual-nh** [comment]
Hrm. I'm going to need to put in a clear_avatar() before the initial debug_weary_info. I had meant this to help tell what was happening just before, but it's getting affected by differing heights (and probably ages). This will be temporary, since it is redundant with do_activity having clear_avatar (unless do_activity has a parameter added for whether to clear the avatar; not sure how acceptable this would be).

**Comment by actual-nh** [comment]
OK, with that done, there are no differences from different seeds in the default order (which is not dictated by the filter order in a file, BTW):
[weary3_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898917/weary3_log.txt)
[weary2_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898918/weary2_log.txt)
[weary1_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898919/weary1_log.txt)

Let's see if changing the order, but still only running the weary tests, will do things... yes:
[weary1_orderR_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898938/weary1_orderR_log.txt)
[weary2_orderR_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898939/weary2_orderR_log.txt)
[weary3_orderR_log.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/5898942/weary3_orderR_log.txt)

**Comment by actual-nh** [comment]
The Mac build (in the General build matrix) failed when the crafting_skill_gain test process aborted:
`Assertion failed: (test_world != nullptr), function init_global_game_state, file ../tests/test_main.cpp, line 130.`
I somehow doubt this is due to my modifications, particularly since they work fine locally. @BrettDong, if you or someone else on a 10.15+ machine could check out the uploaded cata_test? (I have to wonder whether there is some unexpected interaction - file access/writing most likely - between the different test processes.)

**Comment by BrettDong** [comment]
I also noticed intermittent `Assertion failed: (test_world != nullptr)` error on other PR checks. The error can disappear after rerun. It feels like another mysterious bug. I have never encountered such error on my Mac (M1, macOS 11.1).

**Comment by BrettDong** [comment]
Your guess is correct. I managed to reproduce this error if I run tests in parallel like `build-scripts/build.sh` does. I'm looking into this.

**Comment by BrettDong** [comment]
Should be fixed in #47134.

**Comment by actual-nh** [comment]
I am now somewhat debating whether to remove the added `clear_avatar()` calls (and the `debug_weary_info()` calls they make useful); they are otherwise redundant with the same call done in `do_activity`. But they can give useful information, and there are still problems with weariness according to various of the weary tests (both interaction with other tests and the problematic fluctuation in weary levels during various tests - I may actually add tests to check for those). Any thoughts?

**Comment by anothersimulacrum** [comment]
If it gives more information that may help diagnose a test failure, I don't see reason to strip it out.

**Comment by actual-nh** [comment]
> If it gives more information that may help diagnose a test failure, I don't see reason to strip it out.

Well, the Travis tests _finally_ completed with no errors; I think this is ready.

File: tests/player_helpers.cpp [hunk 1]
```diff
@@ -72,6 +72,8 @@ void clear_character( player &dummy )
     // This sets HP to max, clears addictions and morale,
     // and sets hunger, thirst, fatigue and such to zero
     dummy.environmental_revert_effect();
```
Comment (actual-nh) [inline tests/player_helpers.cpp:74]: Anyone have any idea whether this originally intended to reset calories along with hunger "and such"?

**Pull request merged**